src/java/main/org/apache/zookeeper/server/ZooKeeperServer.java

```diff
@@ -486,19 +475,57 @@ protected void startSessionTracker() {
         ((SessionTrackerImpl)sessionTracker).start();
     }
 
+    /**
+     * Sets the state of ZooKeeper server. After changing the state, this will
+     * notify the server state changes to registered shutdown handler, if any.
+     * <p>
+     * Following are the server state transitions:
+     * <li>During startup the server will be in INITIAL state.</li>
+     * <li>After the successful startup the server will set the state to
+     * RUNNING.</li>
+     * <li>Server will enter into ERROR state if it hits an internal error.
+     * {@link ZooKeeperServerListenerImpl} will notify critical resource error
+     * events.</li>
+     * <li>During shutdown the server will set the state to SHUTDOWN, which
+     * represents that the server is not running.</li>
+     *
+     * @param state new server state.
+     */
+    void setState(State state) {
+        this.state = state;
+        // Notify server state changes to the registered shutdown handler, if any.
+        if (zkShutdownHandler != null) {
```

Comment:
Yes, zkServer.setState() will be called inside #notifyStopping(). I'm changing #setState to protected because org.apache.zookeeper.server.quorum.QuorumZooKeeperServer and org.apache.zookeeper.server.ZooKeeperServer classes are in different package. Presently, #setState() is having default package visibility so for overriding #setState in subclass, it needs to increase the visibility, right? Same reason for "enum State" protected.